src/utilities/geometry/Polyhedron.cpp

```diff
@@ -198,72 +317,161 @@ Polyhedron Polyhedron::updateZonePolygonsForMissingColinearPoints() const {
+  edgesNotTwo.reserve(numVertices());
+
+  // All edges for an enclosed polyhedron should be shared by two (and only two) side
+  for (const auto& surface : m_surfaces) {
+    for (const Surface3dEdge& thisSurface3dEdge : surface.edges) {
+      // only return a list of those edges that appear in both the original edge and the revised edges:
+      // this eliminates added edges that will confuse users (edges that were caught by the updateZonePolygonsForMissingColinearPoints routine)
+      if ((thisSurface3dEdge.count() != 2) && (includeCreatedEdges || !thisSurface3dEdge.hasBeenCreated())) {
+        auto itFound = std::find(edgesNotTwo.begin(), edgesNotTwo.end(), thisSurface3dEdge);
+        if (itFound == edgesNotTwo.end()) {
+          edgesNotTwo.push_back(thisSurface3dEdge);
+        }
+      }
+    }
+  }
+  for (auto& edge : edgesNotTwo) {
+    LOG(Debug, edge);
+  }
+
+  return edgesNotTwo;
+}
+
+boost::optional<Surface3dEdge> Surface3dEdge::splitEdge(Point3d testVertex) {
+  if (containsPoint(testVertex)) {
+    Surface3dEdge newEdge(*this);
+    this->m_end = testVertex;
+    newEdge.m_start = std::move(testVertex);
+    this->markCreated();
+    newEdge.markCreated();
+    return newEdge;
+  }
+
+  return boost::none;
+}
+
+std::vector<Surface3d> Polyhedron::findSurfacesWithIncorrectOrientation() const {
+
+  if (!m_hasAnySurfaceWithIncorrectOrientation) {
+    return {};
+  }
+
+  if (m_isCompletelyInsideOut) {
+    LOG(Error, "It seems that ALL surfaces are reversed");
+    return m_surfaces;
+  }
+
+  if (!m_isEnclosedVolume) {
+    LOG(Warn, "Polyhedron is not enclosed. Looking surfaces with incorrect orientations can return false negatives.");
+    //  return {};
+  }
+
+  std::vector<Surface3dEdge> uniqueSurface3dEdges = uniqueEdges();
+
+  // Remove non-conflicted edges
+  uniqueSurface3dEdges.erase(
+    std::remove_if(uniqueSurface3dEdges.begin(), uniqueSurface3dEdges.end(), [](const auto& edge) { return !edge.hasConflictedOrientation(); }),
+    uniqueSurface3dEdges.end());
+  if (uniqueSurface3dEdges.empty()) {
+    return {};
+  }
+
+  std::set<Surface3d> conflictedSurfaces;
+  for (auto& edge : uniqueSurface3dEdges) {
+    const auto& surfNums = edge.allSurfNums();
+    const size_t sfIndex1 = surfNums.front();
+    const size_t sfIndex2 = surfNums.back();
+
+    const auto& sf1 = m_surfaces.at(sfIndex1);
+    const auto& sf2 = m_surfaces.at(sfIndex2);
+
+    const auto c1 = sf1.ratioOfConflictedEdges();
+    const auto c2 = sf2.ratioOfConflictedEdges();
+
+    LOG(Trace, fmt::format("'{}' has {:.2f}% conflicted edges ({}/{}), while '{}' has {:.2f}% conflicted edges ({}/{})", sf1.name, c1,
+                           sf1.numConflictedEdges(), sf1.edges.size(), sf2.name, c2, sf2.numConflictedEdges(), sf2.edges.size()));
+    if (c1 > c2) {
+      conflictedSurfaces.insert(sf1);
+    } else {
+      conflictedSurfaces.insert(sf2);
+    }
+  }
+
+  return {conflictedSurfaces.begin(), conflictedSurfaces.end()};
```

Comment:
When determining surfaces with incorrect orientation, we check each conflicted edge.
For each of the two surfaces that share that edge, we compute the ratio of conflicted edges / total number of edges, and we keep the one with the largest ratio as this is probably the one that's in the wrong order.